x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/CheckTargetShardsCountStep.java

```diff
@@ -69,4 +63,22 @@ public Result isConditionMet(Index index, ClusterState clusterState) {
 
         return new Result(true, null);
     }
+
+    private Integer getTargetNumberOfShards(String policyName, ClusterState clusterState) {
+        IndexLifecycleMetadata indexLifecycleMetadata = clusterState.metadata().custom(IndexLifecycleMetadata.TYPE);
+        LifecycleAction lifecycleAction = indexLifecycleMetadata.getPolicyMetadatas()
```

Comment:
More of a NIT: Should we be a little more careful here to avoid NPEs if the policy got concurrently modified/deleted?

Reply:
Hm, good point. We should, but then we need to see how to handle it. I do not have a clear picture yet how we handle the deletion or if the step got removed. Do I remember correctly that keep a cached version of the step being executed in the index metadata? Should we fallback there?

Reply:
I think throwing an appropriate exception that explains what happened is just fine here. No need to add additional logic, just figured it'd be nice to avoid a possible NPE and replace it with an easy to interpret exception.

Reply:
I see, I am still doubting if that is sufficient though. What concerns me in this approach is that it will throw errors and get stuck there permanently. That will also change the code behavior compared to what it does now. 
If this would have happened in the current set-up, the check would yield either true or false and it would continue. That's what I am trying to achieve by looking for a graceful fallback. Don't you think that's worth it?

Reply:
That's a fair point if we get stuck in error permanently. I thought other functionality would take care of clearing things up the policy disappears? (maybe that's something to fix separately if not?)
That said, that's a different issue from what we're working on here IMO. Here I was just aiming at getting an easy to understand exception.

> the check would yield either true or false and it would continue. 

Right but then we'd throw on the next step anyway (at least we should) wouldn't we so likely not really a change from the user's perspective?

Reply:
True, I need to check what happens now and how can we nicely deal with it, if it's an error in this step or the next then definitely an error is sufficient. If there is a way of handling the deleted/changed policy during a step transition, then we need to not fail and move to that transition. @andreidan can you shed some light?

Reply:
We don't allow deleting a policy whilst it's used by any index. 
Equally, steps execute based on the cached phase. The cached phase is refreshed if it's "safe" (safe currently means the number of steps in the new phase == number of steps in the cached phase - so the action cannot disappear from underneath our feet https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/PhaseCacheManagement.java#L243)

Reply:
Thanks for looking into this failure Mary

I think this will potentially break our contract towards the phase we cache for execution.

For e.g. let's say an index is in the `warm` phase that currently has actions `shrink to 3 shards and forcemerge`. Our index is in the `shrink` action. 
A user updates the `warm` phase to `shrink to 1 shard` (also removing `forcemerge`). Since we removed an action, all the indices currently in the `warm` phase MUST execute the **previously defined** (and cached) phase - to honour the started flow https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/PhaseCacheManagement.java#L243
So an index in the `warm` phase will continue to execute `shrink to 3 and forcemerge`.

With _this_ change, the user will pick up a `shrink` action to a different target (1 shard) and the `forcemerge` will also be executed (as we didn't update the cached phase). 

I don't think we should make this change, but rather rework the test.

Reply:
@andreidan if I understand correctly then the safest option that fixes also the outdated step count issue it to read it from the cached phase, right?
Assuming that if you update only the target shard count then the cached phase will be updated because the number of  steps remained the same.

Reply:
++ that's a safe option

Reply:
What if we combine the two approaches to be a bit more efficient. What if we compare if the number of shards in the cluster state and in the step are the same (that is probably the most common case). If they are different then we try to parse the cached phase in an effort to improve user experience and we accept the performance hit since this will probably will not be a very common situation?

Reply:
@gmarouli That's an interesting idea. 

Do you think this is worth doing though? The `ClusterStateWaitStep` will eventually get the correct value and this affects concurrent updates only. 

IMO we should fix the test and trigger a cluster state update if the test fails just to make sure it's not run into this race condition.

Reply:
You are probably right..... It's probably not worth the trouble... I wanted to explore all the options before I change the test instead for the race condition.